network/src/bounded_executor.rs

```diff
@@ -81,24 +112,107 @@ impl BoundedExecutor {
         F::Output: Send + 'static,
     {
         // Release the permit back to the semaphore when this task completes.
-        let f = f.map(move |ret| {
+        let f = Self::with_permit(f, spawn_permit);
+        self.executor.spawn(f)
+    }
+
+    // Returns a [`Future`] that complies with the `BoundedExecutor`. Once launched,
+    // will block if the executor is at capacity, until one of the other spawned
+    // futures completes.
+    async fn run_on_semaphore<F>(semaphore: Arc<Semaphore>, f: F) -> F::Output
+    where
+        F: Future + Send + 'static,
+        F::Output: Send + 'static,
+    {
+        let permit = Self::acquire_permit(semaphore.clone()).await;
+        Self::with_permit(f, permit).await
+    }
+
+    /// Spawn a [`Future`] on the `BoundedExecutor`. This function is async and
+    /// will block if the executor is at capacity until one of the other spawned
+    /// futures completes.
+    /// In case the future completes with an error, it retries, queueing attempts
+    /// on the semaphore according to the provided [`crate::RetryConfig`].
+    /// This function returns a [`JoinHandle`] that the caller can `.await` on for
+    /// the results of the overall set of attempts to execute the [`Future`].
+    pub async fn spawn_with_retries<F, Fut, T, E>(
+        &self,
+        retry_config: crate::RetryConfig,
+        mut f: F,
+    ) -> JoinHandle<Result<T, E>>
+    where
+        F: FnMut() -> Fut + Send + 'static,
+        Fut: Future<Output = Result<T, backoff::Error<E>>> + Send + 'static,
+        T: Send + 'static,
+        E: Send + 'static,
+    {
+        let retrier = || {
+            let semaphore = self.semaphore.clone();
+
+            let executor = move || {
+                let semaphore = semaphore.clone();
+                BoundedExecutor::run_on_semaphore(semaphore, f())
+            };
+
+            retry_config.retry(executor)
+        };
+        self.executor.spawn(retrier())
```

Comment:
This function does not do what the documentation states it does. In particular this function, while marked as async, does not contain any awaits and will unconditionally spawn a future onto the runtime, even if we're at capacity.

Reply:
It will indeed unconditionally spawn a task on the runtime that queues a sequence of sends, one after the other, each of which are semaphore-bound.

What else would you like to have happen?

Reply:
I would like the doc comment on the function to correctly describe what is going on in this function. Also given the current structure it could be made a non-async function

Reply:
I also have trouble understanding this function.

My understanding is that `self.executor.spawn(retrier())` will do the following:
- it will first execute the closure above it. So this does not pass the closure as an argument but its value after execution.
- then passes the return value of the closure to the `self.executor.spawn` function (which I have to say I am not entirely familiar with).

Reply:
it executes the closure, which returns a future to be executed - you can't restart a future from the beginning so this is necessary

Reply:
I've removed the useless eta-expansion (which was just useful for me to figure out lifetimes & scoping)